### ISSUE TITLE
[WASM-Export] Unable to use multiple rust GDExtensions at the same time

### ISSUE THREAD

**Issue by Guusggg** [godot-rust/gdext#968]
Hi,

Lately I've been working on a bigger game, and completed the refactor from C# to C++ and then I found this amazing library so I switched one more time to Rust. 

Now, what I've had in mind was that I'd create a Shared, Client and Server project in Rust. Shared would have the Nodes that both projects need and Client and Server would be implementing the multiplayer features. 

I created the following setup:
- godot (Godot Project)
- rust (Rust workspace)
    - client (Rust project)
    - server (Rust project)
    - shared (Rust project)

Note: The `client` nor `server` projects do not depend on `shared` in `Cargo.toml`.

This setup is working terrific! I'm able to implement client specific features, and server specific features which will be connected using the shared nodes. 

I can run this on my Linux machine, and confirm that it is working. However, I also target `wasm` so I tried getting that to work. Now, this isn't really working out. 

I compile the workspace with `cargo +nightly build -Zbuild-std --target wasm32-unknown-emscripten` to WebAssembly, and this is going really well. But when I run the project, on Firefox I get an `out of memory` emscripten error (which it also does with a single Rust project, but only on Debug builds). When I run it on Chromium however, I get a big list of errors:
```
RuntimeError: Aborted(undefined). Build with -sASSERTIONS for more info.
    at abort (tmp_js_export.js:9:8604)
    at ___cxa_throw (tmp_js_export.js:9:425328)
    at shared.wasm.panic_unwind::imp::panic::h7a89a72211144e04 (shared.wasm-10a1508e:0x42b504)
    at shared.wasm.__rust_start_panic (shared.wasm-10a1508e:0x42b367)
    at shared.wasm.rust_panic (shared.wasm-10a1508e:0x41d2e3)
    at shared.wasm.std::panicking::rust_panic_with_hook::hc3a8cd21f77927d3 (shared.wasm-10a1508e:0x41d1d9)
    at shared.wasm.std::panicking::begin_panic_handler::_$u7b$$u7b$closure$u7d$$u7d$::hb19be6fa9857ee61 (shared.wasm-10a1508e:0x41a5f5)
    at shared.wasm.std::sys::backtrace::__rust_end_short_backtrace::h585b6ab617e4600a (shared.wasm-10a1508e:0x41a4bd)
    at shared.wasm.rust_begin_unwind (shared.wasm-10a1508e:0x41c2a1)
    at shared.wasm.core::panicking::panic_fmt::he773db1f33281f18 (shared.wasm-10a1508e:0x49a208)

  | onPrintError | @ | tmp_js_export.js:477
-- | -- | -- | --
  | abort | @ | tmp_js_export.js:9
  | ___cxa_throw | @ | tmp_js_export.js:9
  | $panic_unwind::imp::panic::h7a89a72211144e04 | @ | shared.wasm-10a1508e:0x42b504
  | $__rust_start_panic | @ | shared.wasm-10a1508e:0x42b367
  | $rust_panic | @ | shared.wasm-10a1508e:0x41d2e3
  | $std::panicking::rust_panic_with_hook::hc3a8cd21f77927d3 | @ | shared.wasm-10a1508e:0x41d1d9
  | $std::panicking::begin_panic_handler::_$u7b$$u7b$closure$u7d$$u7d$::hb19be6fa9857ee61 | @ | shared.wasm-10a1508e:0x41a5f5
  | $std::sys::backtrace::__rust_end_short_backtrace::h585b6ab617e4600a | @ | shared.wasm-10a1508e:0x41a4bd
  | $rust_begin_unwind | @ | shared.wasm-10a1508e:0x41c2a1
  | $core::panicking::panic_fmt::he773db1f33281f18 | @ | shared.wasm-10a1508e:0x49a208
  | $godot_ffi::binding::single_threaded::BindingStorage::set_initialized::haf88d1160aa08dd6 | @ | shared.wasm-10a1508e:0x3a7ebd
  | $godot_ffi::binding::single_threaded::BindingStorage::initialize::h06c546cc4398088b | @ | shared.wasm-10a1508e:0x3a7f1e
  | $godot_ffi::binding::initialize_binding::hec449e30c65edce2 | @ | shared.wasm-10a1508e:0x3b4174
  | $godot_ffi::initialize::hd3864843f056cfd2 | @ | shared.wasm-10a1508e:0x3b53dc
  | $godot_core::init::__gdext_load_library::_$u7b$$u7b$closure$u7d$$u7d$::h80e68cd7c7203dc9 | @ | shared.wasm-10a1508e:0x66696
  | $core::ops::function::FnOnce::call_once::h50d84be8fa3c9798 | @ | shared.wasm-10a1508e:0x70136
  | $std::panicking::try::do_call::h64bb311ea799ec2a | @ | shared.wasm-10a1508e:0x741fb
  | $__rust_try | @ | shared.wasm-10a1508e:0x734d8
  | $std::panicking::try::hcbde3b6f2bb9e296 | @ | shared.wasm-10a1508e:0x74088
  | $std::panic::catch_unwind::h4d4f243b6b7cf82f | @ | shared.wasm-10a1508e:0x72b00
  | $godot_core::private::handle_panic_with_print::h28362aa50547f00d | @ | shared.wasm-10a1508e:0x596e6
  | $godot_core::private::handle_panic::h7e38a629003b6b68 | @ | shared.wasm-10a1508e:0x59400
  | $godot_core::init::__gdext_load_library::h65f7301edd9aa8a6 | @ | shared.wasm-10a1508e:0x66521
  | $gdext_rust_init | @ | shared.wasm-10a1508e:0x5616f
  | $func58011 | @ | 0aaeb74a:0x2599747
  | $func58034 | @ | 0aaeb74a:0x259dff6
  | $func51263 | @ | 0aaeb74a:0x2466843
  | $func51313 | @ | 0aaeb74a:0x246c885
  | $func34708 | @ | 0aaeb74a:0x19938e1
  | $func1310 | @ | 0aaeb74a:0x4c7148
  | $func1212 | @ | 0aaeb74a:0x431bc4
  | $_Z14godot_web_mainiPPc | @ | 0aaeb74a:0x2cfd7e
  | __Z14godot_web_mainiPPc | @ | tmp_js_export.js:9
  | $__main_argc_argv | @ | 006438ee:0xa5975
  | callMain | @ | tmp_js_export.js:9
  | (anonymous) | @ | tmp_js_export.js:804
  | (anonymous) | @ | tmp_js_export.js:799
  | Promise.then |   |  
  | start | @ | tmp_js_export.js:778
  | (anonymous) | @ | tmp_js_export.js:837
  | Promise.then |   |  
  | startGame | @ | tmp_js_export.js:836
  | (anonymous) | @ | tmp_js_export.html:181
  | (anonymous)
``` 

I've tried using different entry point names using 
```
struct SharedExtension;

#[gdextension(entry_symbol = gdext_shared_rust_init)]
unsafe impl ExtensionLibrary for SharedExtension {}
```

Which doesn't work. As soon as I remove the other extensions (so I leave `shared.gdextension`) it starts working again.

I use these Emscripten compile flags:
```
[target.wasm32-unknown-emscripten]
rustflags = [
    "-C", "link-args=-sSIDE_MODULE=2",
    "-C", "link-args=-pthread", # was -sUSE_PTHREADS=1 in earlier emscripten versions
    "-C", "target-feature=+atomics,+bulk-memory,+mutable-globals",
    "-Zlink-native-libraries=no",
    "-Cllvm-args=-enable-emscripten-cxx-exceptions=0",
]
```

I also created two projects to recreate the issue:
[single-rust-project.zip](https://github.com/user-attachments/files/18115972/single-rust-project.zip)
[multiple-rust-project.zip](https://github.com/user-attachments/files/18115974/multiple-rust-project.zip)

I'm really eager to help this project move forward, and I would say I'm quite well-versed in WebAssembly, however I have no idea where to get started. I tried enabling `-sASSERTIONS` and `-sASSERTIONS=2` but this gives me a `resolved is not a function error` so no dice...

Any tips are very welcome, and if you need more information please let me know!

Thank you for the amazing work on this library, it's awesome!
Kind regards,

**Comment by Bromeon** [comment]
Thanks a lot for the detailed report and the kind words :slightly_smiling_face: 

> I tried enabling `-sASSERTIONS` and `-sASSERTIONS=2` but this gives me a `resolved is not a function error` so no dice...

Yes, this is a known behavior, I had the same experience ([Discord thread](https://discord.com/channels/723850269347283004/1274080437198651546/1275339473135206441)). It should be correct without those.

---


Looking at your stack trace, problem seems to come from here:

https://github.com/godot-rust/gdext/blob/4bc92e8cbd02c9cef661a7921df5adae7ed55790/godot-ffi/src/binding/single_threaded.rs#L72-L92

I don't know which panic is exactly triggered, is there maybe a way to log/debug the message and/or line number?

Since this is also about threading, one check is to make sure your threading settings are in sync.
You need (mutually exclusive):
- Either:
   - Export options in Godot with thread support
   - Enable `experimental-wasm` feature
   - Emcc config `-C pthread` (like you have now)
- Or:
   - Export options in Godot without thread support
   - Enable **both** `experimental-wasm` and `experimental-wasm-nothreads`
   - Remove the `-C pthread` option

---

Another few points you can try that have caused problems in the past:

1. Do you use the `lazy-function-tables` feature?
2. Can you check the `.wasm` file names of the different libraries, and make sure they're distinct?
(We had an issue with renamed `.wasm` files that weren't found in https://github.com/godot-rust/gdext/pull/799, although not directly related).
3. Someone added other emcc flags:
    ```toml
    rustflags = [
        ...
        "-Clink-arg=-fwasm-exceptions",
        "-Clink-args=-sDISABLE_EXCEPTION_CATCHING=1",
        "-Clink-args=-sEXPORT_ALL=1",
        "-Clink-args=-sSUPPORT_LONGJMP=wasm",
        "-Cllvm-args=-enable-emscripten-cxx-exceptions=0",
        "-Cllvm-args=-wasm-enable-sjlj",
    ]
    ```
4. Emscripten version: I personally couldn't find any difference, but some people tested 3.1.39 last year. For me things worked with latest (but my setup was simple).

There is also quite a bit of knowledge in https://github.com/godot-rust/gdext/issues/438. Unfortunately it's still quite scattered, there's some work to be done in improving the web-export page in the book.

**Comment by Guusggg** [comment]
Hi, 

Thank you so much for the quick reply and the helpful links! 

>Export options in Godot with thread support
>Enable experimental-wasm feature
>Emcc config -C pthread (like you have now)

This is how I've been doing it when I submitted the projects, so I guess this isn't it.

>Export options in Godot without thread support
>Enable both experimental-wasm and experimental-wasm-nothreads
>Remove the -C pthread option

This also causes the same error, however, exporting using Release now gives me the error: `already initialized`. So I guess where are somewhere on track here.

For both approaches, removing the second `gdextension` makes it run again (although with errors about missing classes, obviously).

1. Currently I'm not using the `"lazy-function-tables"` option, but I have tried it in the past to get faster compile times.
2. In my export folder, I see `client.wasm` and `shared.wasm` as (should be right?) expected.
3. With all of these, same thing. Unable to load the program. I've tried with both `-pthread` and without.

```
Aborted()
onPrintError	@	index.js:474
abort	@	index.js:9
__abort_js	@	index.js:9
$abort	@	0062dbd6:0xa3fc5
$std::sys::pal::unix::abort_internal::h5ffb198ad2b8db0d	@	client.wasm-0f92f976:0x3c3083
$std::panicking::rust_panic_with_hook::hc3a8cd21f77927d3	@	client.wasm-0f92f976:0x3ce88c
$std::panicking::begin_panic_handler::_$u7b$$u7b$closure$u7d$$u7d$::hb19be6fa9857ee61	@	client.wasm-0f92f976:0x3cbbe9
$std::sys::backtrace::__rust_end_short_backtrace::h585b6ab617e4600a	@	client.wasm-0f92f976:0x3cbab1
$rust_begin_unwind	@	client.wasm-0f92f976:0x3cd895
$core::panicking::panic_nounwind_fmt::runtime::h9938091b05990098	@	client.wasm-0f92f976:0x44c31b
$core::panicking::panic_nounwind_fmt::h6e7e6d2ffc89c5c5	@	client.wasm-0f92f976:0x44c2b2
$core::panicking::panic_nounwind::h99d9192d3d1b4d38	@	client.wasm-0f92f976:0x44c3c4

$core::ptr::replace::precondition_check::h75dfa671cd82b880	@	client.wasm-0f92f976:0x3c1167
$core::ptr::replace::h9395aa21c0716916	@	client.wasm-0f92f976:0x3c12dd
$std::sys::thread_local::native::lazy::Storage$LT$T$C$D$GT$::initialize::h4975e865484cb4ea	@	client.wasm-0f92f976:0x3a0094
$std::sys::thread_local::native::lazy::Storage$LT$T$C$D$GT$::get_or_init::hd146c2b9dcadca57	@	client.wasm-0f92f976:0x39fe95
$std::hash::random::RandomState::new::KEYS::_$u7b$$u7b$constant$u7d$$u7d$::_$u7b$$u7b$closure$u7d$$u7d$::h259f534f630663f1	@	client.wasm-0f92f976:0x3a036d
$core::ops::function::FnOnce::call_once::hf0fb143070654117	@	client.wasm-0f92f976:0x35457c
$std::thread::local::LocalKey$LT$T$GT$::try_with::h531f46e8c61644d2	@	client.wasm-0f92f976:0x35de7f
$std::thread::local::LocalKey$LT$T$GT$::with::h366e74810743984b	@	client.wasm-0f92f976:0x35ddc8
$std::hash::random::RandomState::new::h5ad2fb59cf112d55	@	client.wasm-0f92f976:0x353b3e
$_$LT$std..hash..random..RandomState$u20$as$u20$core..default..Default$GT$::default::h8ff89411a43748c1	@	client.wasm-0f92f976:0x35b3b6
$_$LT$std..collections..hash..map..HashMap$LT$K$C$V$C$S$GT$$u20$as$u20$core..default..Default$GT$::default::hd24fc6464a267ba0	@	client.wasm-0f92f976:0x35339f
$std::collections::hash::map::HashMap$LT$K$C$V$GT$::new::h1f4c9b7444ff5bb3	@	client.wasm-0f92f976:0x35336e

$godot_ffi::string_cache::StringCache::new::h62ad0e819dbe3e71	@	client.wasm-0f92f976:0x364053

$godot_ffi::initialize::hc53ec13c08f5642d	@	client.wasm-0f92f976:0x35d3d3
$godot_core::init::__gdext_load_library::_$u7b$$u7b$closure$u7d$$u7d$::h91332e9a4c846654	@	client.wasm-0f92f976:0x60b07
$core::ops::function::FnOnce::call_once::h64128a8d9be8ca9e	@	client.wasm-0f92f976:0x5e867
$std::panicking::try::do_call::h8565294c672b6171	@	client.wasm-0f92f976:0x68466
$__rust_try	@	client.wasm-0f92f976:0x681fc
$std::panicking::try::h129b92da609d815c	@	client.wasm-0f92f976:0x682f3
$std::panic::catch_unwind::hdd122ce3867bdf42	@	client.wasm-0f92f976:0x4ea98
$godot_core::private::handle_panic_with_print::h6d40b7b7e0423e98	@	client.wasm-0f92f976:0x517f1
$godot_core::private::handle_panic::h2142ae78a69be744	@	client.wasm-0f92f976:0x51517
$godot_core::init::__gdext_load_library::hb155d0c815240a5c	@	client.wasm-0f92f976:0x609db

$gdext_rust_init
``` 

A couple things that I find interesting in the stacktrace is:
- `$core::ptr::replace::precondition_check::h75dfa671cd82b880	@	client.wasm-0f92f976:0x3c1167` This is the last function being called before something with `unwind`, it's also suspicious that it has `_check` in the name. Could this be triggering the actual panic?
- `$godot_ffi::string_cache::StringCache::new::h62ad0e819dbe3e71	@	client.wasm-0f92f976:0x364053` This is the last function being called from code that is coming from this repository. Could it be that WebAssembly has a hard time with "random" things? I know using WASM before in Rust that people recommended using the Javascript functions for random as it is not in `std` in Rust.

I'm browsing the source code of the `godot-rust` crates a bit, to see if there's something that I think that could cause a panic. I'm not sure how to get a better stacktrace, but I did see `debug-assertions` in the `godot-core` crate. Is there a way to turn those on or rather, is it as simple as adding "debug_assertions" to features?

Kind regards,

**Comment by Guusggg** [comment]
Thank you so much again for the quick reply! You got me tinkering...!

I'm unable to finish the research tonight but I've learned quite a bit already. For one thing, I'm almost certain that `gdext` is being initialized twice, as when I compile it with `"__debug-log"`, I see the message `Initialize gdext...` twice.

I don't really understand the impact of this yet, but I'm sure it's not a good thing :P. I suspect it certainly has to do something with the errors I'm getting, now to find the culprit of how to fix it. I'm sure I don't have enough knowledge to really fix it, but I hope I could make a proposal and with you're help, refine it into a nice fix.

Thanks again for your work, I really appreciate it. I'll see if I can get a hacky fix, and I'll let you know. I'd be really happy if you'd be able to help me along then too!

**Comment by Bromeon** [comment]
I think a proper fix would be to add an attribute to the `gdextension` proc macro:
```rs
#[gdextension(dependency)]
unsafe impl ExtensionLibrary for ClientExtension {}
```

And based on that, certain initialization code would not be generated and/or not run.

In case you want to tinker, you could search for `handle_alone` in the codebase. That allows you to detect the key `dependency` in the above macro :slightly_smiling_face:

**Comment by Guusggg** [comment]
@Bromeon Hi, I created a pullrequest at #973. I know it's not of high quality, and I was just looking to get it to work. 

Do not merge it! :P

Could you take a look? I've encountered some strange things that I'll be detailing in the pull request.

**Comment by SvizelPritula** [comment]
> I think a proper fix would be to add an attribute to the `gdextension` proc macro:

I don't think that makes much sense. As far as I can tell, this issue doesn't have much to do with dependencies at all. Currently, it doesn't seem to be possible to use two extensions using `gdext` in web builds at all, even if they are completely unrelated. In fact, the extensions in the example project @Guusggg provided don't depend on each other at all. I was able to reproduce the same issues today with two toy extensions of my own, I can provide the project if you want.

This is not a big issue for people who with to write their game in Rust, as they can just keep all code inside one library. However, if someone wants to publish, well, an extension, then this is a huge problem. With `godot-cpp`, you can (I think) just compile your extension and distribute it in the Godot Asset Library, and other devs will be able to use it with zero setup. You can't do the same with `gdext` (if you want wasm support), since web builds would break as soon as someone adds a second library using `gdext` to their game. The only solutions seems to be to distribute the library as source code, and guide users through making and building their own extension that combines all Rust libraries they want to use, which is a lot of friction, given all the cross-compilers involved.

I'm not familiar with the internals of `gdext`, but I imagine it must be possible to initialize the library only in the extension that gets loaded first? That must be what happens on other platforms?

**Comment by Yarwin** [comment]
I'll look into it this weekend

**Comment by Yarwin** [comment]
ok, I'm puzzled:

1. I've downloaded the test project ([multiple-rust-project.zip](https://github.com/user-attachments/files/18115974/multiple-rust-project.zip))
2. I specified latest stable godot-rust version for both of them:
```toml
[dependencies]
godot = { version = "0.4.5", features = ["api-custom", "experimental-wasm", "lazy-function-tables"] }
```
3. I've edited .cargo/config.toml file:
```toml
[env]
GODOT4_BIN = ....

[target.wasm32-unknown-emscripten]
rustflags = [
    "-C", "link-args=-pthread", 
    "-C", "target-feature=+atomics", 
    "-C", "link-args=-sSIDE_MODULE=2",
    "-C", "link-args=-fno-entry=1",
    "-Zlink-native-libraries=no",
    "-Cllvm-args=-enable-emscripten-cxx-exceptions=0",
    "-Z", "emscripten-wasm-eh=false",
]
```

And... Everything seems to work fine?

<img width="1700" height="585" alt="Image" src="https://github.com/user-attachments/assets/8ad8669e-dba8-46b8-b76e-68c074319a60" />

Note that one needs to check "threads" (export -> web -> thread support) to allow loading multiple extensions which might share the same memory: https://github.com/emscripten-core/emscripten/issues/12098

@SvizelPritula could you please provide some MRP, similar to the linked one?

**Comment by SvizelPritula** [comment]
I originally didn't think to try a threaded export, but it doesn't seem to work with threads either, it panics with:

```
playground.js:488 ERROR: [panic /home/dev/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/godot-core-0.4.5/src/meta/class_id.rs:266]  insert_class_name() called for already-existing string: RefCounted
```

The project is here: [playground.zip](https://github.com/user-attachments/files/24421738/playground.zip). The build of each of the two extensions is governed by a Makefile, which builds all the required libraries for x86_64 Linux and WebAssembly.

**Comment by Yarwin** [comment]
Thank you! I can reproduce the same problem on my side.

**Comment by Yarwin** [comment]
It seems that symbols/globals defined in a `godot` crate point to the very same place/instance in the memory no matter from what crate they are being called from. I haven't checked generated `.wasm` binary but I did check it empirically: 

<img width="547" height="369" alt="Image" src="https://github.com/user-attachments/assets/22bedc9c-0cb0-40a5-ac1e-08eac7a4d6af" />

In this example independent crate A (sincalculator) and independent crate B (coscalculator) are referring to the very same `ClassIdCache`. 

As far as I'm aware it is case for other symbols as well. This might lead to instant UB.



We need to create MRP and hit the upstream (since I'm unsure which one, I would start with rust forums).

**Comment by Bromeon** [comment]
I read up (very surface-level) on Wasm + Emscripten model a bit. What I found:
- Wasm has "main modules" + "side modules" that effectively share a single runtime, sharing memory.
- Emscripten typically uses 1 `MAIN_MODULE` (JS environment + system libraries) + multiple `SIDE_MODULE`s (shared libraries, pure Wasm). This is also the Godot setup, necessary so that extensions can communicate with each other and the engine.
- Dynamic linking in our case compiles to a side module. [We use `link-args=-sSIDE_MODULE=2` flag for the compiler.](https://godot-rust.github.io/book/toolchain/export-web.html?highlight=config.toml#project-configuration)
- There seems to be a situation in which global variables are resolved only once across multiple side modules. I need further research to fully understand why (are 2 `godot` crates combined to 1 runtime instance? are global variables deduplicated/symbol-merged if they have the same name?)

We need to find a way to isolate globals from multiple `godot` crates. At first I thought [`-sMODULARIZE`](https://emscripten.org/docs/compiling/Modularized-Output.html) could be used for isolation, but that seems to be only for the main module and JS code, solving a different problem.

So ideally we find a way to configure compiler/linker to emit isolated memory in the side modules. If we do not achieve that, we need to rework how _all_ global variables work in our libraries, e.g. by having a runtime key based on the extension name/hash. Such a thing could maybe be abstracted in `Global<T>`.

I do wonder if [`-Zdefault-visibility=hidden`](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/default-visibility.html) could help us avoid "interposing" of global variables.

Potentially interesting:
- https://emscripten.org/docs/compiling/Dynamic-Linking.html
- https://developer.mozilla.org/en-US/docs/WebAssembly/Reference/JavaScript_interface/Memory

**Comment by SvizelPritula** [comment]
Here's what I've been able to figure out:

First of all, the problem occurs at compilation time. The [`CLASS_ID_CACHE`](https://github.com/godot-rust/gdext/blob/95791a4cc35640c18dbba1ec3fa5c17058b819b9/godot-core/src/meta/class_id.rs#L28) `static` variable, along with several others, gets turned into an import:
```wasm
  (import "GOT.mem" "_RNvNtNtCsiZxTL9DORka_10godot_core4meta8class_id14CLASS_ID_CACHE" (global (;22;) (mut i32)))
```

This means that it [will be given an address by Emscripten's dynamic linker at runtime](https://github.com/WebAssembly/tool-conventions/blob/8e3191e4992b7e96369ebcfec3af86610464ec27/DynamicLinking.md#imports). All modules (so GDExtensions) that import a symbol with the same name will be given the same address. Therefore, all extensions using `godot-rust` will share the same `CLASS_ID_CACHE`. This will cause the crate to panic, which is probably for the best, as some other `static`s, such as [`ERROR_PRINT_LEVEL`](https://github.com/godot-rust/gdext/blob/95791a4cc35640c18dbba1ec3fa5c17058b819b9/godot-core/src/private.rs#L49), don't get shared between extensions. This means that neither Godot or Emscripten's runtime are probably at fault.

Now, the main question is: Why does Cargo output a binary that handles some global variables this way?

Here's my summary of relevant linking terminology. I hope I got everything right, I only learned it today. Feel free to correct me.

> In LLVM, each symbol has a [linkage type](https://llvm.org/docs/LangRef.html#linkage-types) and a [visibility style](https://llvm.org/docs/LangRef.html#visibility-styles).
> 
> Linkage types mostly apply to static linking, but also impact dynamic linking. There are two linkage types that are relevant to this discussion:
> 
> - `internal` - Only accessible to the current compilation unit, cannot collide with symbols from other compilation units. Corresponds to the `LOCAL` binding type in ELF files.
> - `external` (the default) - Can be accessed from different compilation units, its name must be unique within the "final" binary. Corresponds to the `GLOBAL` binding type in ELF files.
> 
> Visibility styles are only really relevant for shared libraries. There are two relevant visibility styles:
> 
> - `hidden` - Not visible outside of the shared library, cannot collide with symbols from other libraries. Called `HIDDEN` in ELF files.
> - `interposable`, aka `default` - Visible outside of the shared library, can be overwritten by shared libraries loaded earlier. Called `DEFAULT` in ELF files.

Suppose we have an `rlib` (i.e. not the top level `cdylib`) crate that contains two statics, one public, one private:
```rs
pub static PUBLIC: AtomicI64 = AtomicI64::new(1);
static PRIVATE: AtomicI64 = AtomicI64::new(1);

// Public function using both statics, to keep them alive.
#[inline(never)]
pub fn inc() {
    PUBLIC.fetch_add(1, Ordering::SeqCst);
    PRIVATE.fetch_add(1, Ordering::SeqCst);
}
```

`rustc` will emit the following LLVM IR for it:
```llvm
@_RNvCs6XryrqRON7Q_3dep6PUBLIC = global [8 x i8] c"\01\00\00\00\00\00\00\00", align 8
@_RNvCs6XryrqRON7Q_3dep7PRIVATE = internal global [8 x i8] c"\01\00\00\00\00\00\00\00", align 8
```

This means that `PUBLIC` is `external` and `interposable` and `PRIVATE` is `internal` and `interposable`. Notably, it emits the same IR for both web assembly and x86-64 Linux, aside from the name.

On x86-64 Linux, the symbols that end up in the finished shared library look like this:
```
    31: 00000000000542e0     8 OBJECT  LOCAL  DEFAULT   27 _RNvCs6XryrqRON7Q_3dep7PRIVATE
    34: 00000000000542d8     8 OBJECT  LOCAL  DEFAULT   27 _RNvCs6XryrqRON7Q_3dep6PUBLIC
```

Both symbols are `internal` and `interposable`. Apparently, the previously `external` `PUBLIC` got changed to `internal` either by LLVM or by the linker. This means that each shared library would get its own copy of both globals.

Now, I don't have an easy way to print the attributes of symbols in web assembly, but the generated wasm file contains an import for `PUBLIC` but not `PRIVATE`, which seems to suggest that `PUBLIC` is still `external` and `interposable` and `PRIVATE` is still `internal` and `interposable`.

There are two ways to make each extensions have its own copy of the global variables. The first options is to make sure that no `static`s are public. This is harder than it looks, as `static`s can get `external` linkage even if they're not marked as `pub`, for example if they're used in inlinable functions. A reliably way to force them to be `internal` seems to be to declare them in the body of an uninlinable functions, probably a simple getter. The second and much better option is to change their visibility style, which can be done globally with the unstable [`-Zdefault-visibility=hidden`](https://doc.rust-lang.org/beta/unstable-book/compiler-flags/default-visibility.html).

So, where is the bug? It's likely not in Godot. I'm doubtful that the Rust compiler is at fault, since it generates identical IR for both tested platforms, but maybe that's not what they're supposed to do. That would leave LLVM or Emscripen, or Rust's invocation of either of them. It could also be intended behaviour, but I hope not, as it's very confusing.

**TLDR**: Use `-Zdefault-visibility=hidden`.

**Comment by xiSage** [comment]
test
